qbraid_qir/qasm3/visitor.py

```diff
@@ -133,17 +145,38 @@ def entry_point(self) -> str:
     def finalize(self) -> None:
         self._builder.ret(None)
 
-    def _set_scope(self, scope: Scope) -> None:
-        self._scope = scope
+    def _push_scope(self, scope: dict) -> None:
+        self._scope.append(scope)
+
+    def _pop_scope(self) -> None:
+        if len(self._scope) > 0:
+            self._scope.pop()
 
-    def _in_gate(self) -> bool:
-        return self._scope == Scope.GATE
+        raise IndexError("Scope list is empty, can not pop")
+
+    def _get_scope(self) -> dict:
+        if len(self._scope) > 0:
+            return self._scope[-1]
+
+        raise IndexError("No scopes available to get")
+
+    def _check_in_scope(self, var_name: str) -> bool:
+        try:
+            curr_scope = self._get_scope()
+            return var_name in curr_scope
+        except IndexError:
+            return False
```

Comment:
Now seeing this function, it might be even cleaner to just return `None` inside of `_get_scope` instead of raising an `IndexError` if we're just going to except the `IndexError` here anyways. For example:

```python
def _get_scope(self) -> Optional[dict]:
    if not self._scope:
        return None
    return self._scope[-1]

def _check_in_scope(self, var_name: str) -> bool:
    curr_scope = self._get_scope()
    return curr_scope is not None and var_name in curr_scope
```

It's splitting hairs at this point though, so whatever makes the most sense to you works for me.

Reply:
Well, I think this would also work - 
```python3
def _get_scope(self) -> dict:
        if len(self._scope) == 0:
            raise IndexError("No scopes available to get")
        return self._scope[-1]


    def _check_in_scope(self, var_name: str) -> bool:
        curr_scope = self._get_scope()
        return var_name in curr_scope
```
`curr_scope` would at least be a dict. If no scope, we would be anyway raising an `IndexError`. We'll just need to be careful to always push a `dict` to the scope. Will add some code for that in `_push_scope`